bottlerocket/agents/src/bin/ecs-workload-agent/main.rs

```diff
@@ -0,0 +1,582 @@
+                }
+
+                res
+            }
+        }
+    }
+
+    /// Terminate the test runs. This is a noop for this agent.
+    async fn terminate(&mut self) -> Result<(), Self::E> {
+        Ok(())
+    }
+}
+
+/// Loops waiting for the workload test tasks to complete.
+async fn wait_for_test_completion<I>(
+    ecs_client: &aws_sdk_ecs::Client,
+    cluster_name: &str,
+    task_arns: &[String],
+    info_client: &I,
+) -> Result<TestResults, Error>
+where
+    I: InfoClient,
+{
+    let mut retries: u64 = 0;
+    loop {
+        let mut results = test_results(ecs_client, cluster_name, task_arns).await?;
+        // If tests have all passed or failed, return. Otherwise still in progress so wait and retry.
+        if results.outcome == Outcome::Pass || results.outcome == Outcome::Fail {
+            info_client
+                .send_test_update(results.clone())
+                .await
+                .err()
+                .iter()
+                .for_each(|e| log::error!("Unable to send test update: {}", e));
+            return Ok(results);
+        }
+
+        retries += 1;
+        results.other_info = Some(format!("Waiting for test completion: attempt {}", retries));
+        info_client
+            .send_test_update(results.clone())
+            .await
+            .err()
+            .iter()
+            .for_each(|e| log::error!("Unable to send test update: {}", e));
+        tokio::time::sleep(Duration::from_secs(2)).await;
+    }
+}
+
+/// Gets the status from our test tasks and populates a `TestResults` struct with the outcome.
+async fn test_results(
+    ecs_client: &aws_sdk_ecs::Client,
+    cluster_name: &str,
+    task_arns: &[String],
+) -> Result<TestResults, Error> {
+    let tasks = ecs_client
+        .describe_tasks()
+        .cluster(cluster_name)
+        .set_tasks(Some(task_arns.to_vec()))
+        .send()
+        .await
+        .context(error::TaskDescribeSnafu)?
+        .tasks()
+        .map(|tasks| tasks.to_owned())
+        .context(error::NoTaskSnafu)?;
+    let passed_count = tasks
+        .iter()
+        .filter(|task| task.last_status() == Some("STOPPED"))
+        .filter(|task| task.stop_code() == Some(&TaskStopCode::EssentialContainerExited))
+        .filter(|task| {
+            task.containers()
+                .unwrap_or_default()
+                .iter()
+                .filter(|container| container.exit_code() != Some(0))
+                .count()
+                == 0
+        })
+        .count() as i32;
+    let failed_count = tasks
+        .iter()
+        .filter(|task| task.last_status() == Some("STOPPED"))
+        .filter(|task| task.stop_code() == Some(&TaskStopCode::EssentialContainerExited))
+        .filter(|task| {
+            task.containers()
+                .unwrap_or_default()
+                .iter()
+                .filter(|container| container.exit_code() == Some(0))
+                .count()
+                == 0
```

Comment:
Should it error if any containers exit non-zero instead of requiring all of them?
```suggestion
                .filter(|container| container.exit_code() != Some(0))
                .count()
                != 0
```

Reply:
Yeah, still getting used to these long filter calls and was breaking my brain. I see what you're saying now - will update.